scripts/lib/upgrade-zulip-stage-2

```diff
@@ -295,109 +355,137 @@ else:
-if classes != new_classes:
-    logging.info("Adjusting Puppet classes for renames...")
+# If we're skipping tornado restarts, we need to ensure we keep the
+# same cache prefix as we currently have.
+if args.skip_tornado:
+    shutil.copyfile("../current/var/remote_cache_prefix", "./var/remote_cache_prefix")
+
+# If there are no migrations needed, we can fill the caches now,
+# instead of after they run.
+if not migrations_needed:
     subprocess.check_call(
-        [
-            "crudini",
-            "--set",
-            "/etc/zulip/zulip.conf",
-            "machine",
-            "puppet_classes",
-            ", ".join(new_classes),
-        ]
+        ["./manage.py", "fill_memcached_caches", "--skip-checks"], preexec_fn=su_to_zulip
     )
 
+# If we are planning on running puppet, we can pre-run it in --noop
+# mode and see if it will actually make any changes; if not, we can
+# skip it during the upgrade.  We omit this check if the server is
+# already stopped, since it's not better than just pressing on.
+has_puppet_changes = True
+if not args.skip_puppet and IS_SERVER_UP:
+    logging.info("Pre-checking for puppet changes...")
+    try_puppet = subprocess.run(
+        ["./scripts/zulip-puppet-apply", "--noop", "--force"], stdout=subprocess.DEVNULL
+    )
+    if try_puppet.returncode == 0:
+        if minimal_change:
+            logging.info("Verified no Puppet changes are necessary.")
+        else:
+            logging.info("No puppet changes found, skipping!")
+        args.skip_puppet = True
+        has_puppet_changes = False
+    elif minimal_change:
+        logging.error("Would need to apply puppet changes -- aborting!")
+        sys.exit(1)
 
-if not args.skip_puppet:
-    # Puppet may adjust random services; to minimize risk of issues
-    # due to inconsistent state, we shut down the server first.
-    shutdown_server()
-    logging.info("Applying Puppet changes...")
-    subprocess.check_call(["./scripts/zulip-puppet-apply", "--force"])
-    subprocess.check_call(["apt-get", "-y", "upgrade"])
-    # Puppet may have reloaded supervisor, and in so doing started
-    # services; mark as potentially needing to stop the server.
-    IS_SERVER_UP = True
-
-if migrations_needed:
-    # Database migrations assume that they run on a database in
-    # quiesced state.
-    shutdown_server()
-    logging.info("Applying database migrations...")
-    subprocess.check_call(["./manage.py", "migrate", "--noinput"], preexec_fn=su_to_zulip)
-
-logging.info("Restarting Zulip...")
-if IS_SERVER_UP:
-    restart_args = ["--fill-cache"]
-    if args.skip_tornado:
-        restart_args.append("--skip-tornado")
-    if args.less_graceful:
-        restart_args.append("--less-graceful")
-    subprocess.check_call(["./scripts/restart-server", *restart_args], preexec_fn=su_to_zulip)
+
+if args.skip_restart:
+    logging.info("Successfully configured in %s!", deploy_path)
```

Comment:
Should this print `/home/zulip/deployments/next/scripts/restart-server --fill-cache` as the step to manually restart into the new version, if desired?

Reply:
Yeah, I can envision that `--skip-restart` might take optional values for things that it skips but defers to later.  But I'd rather wait to add those until we actually find ourselves needing them -- as noted below, there's a bunch of complexity here already.